File: Java/257-BinaryTreePaths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> result = new ArrayList<>();
        if (root == null)   return result;
        String s = root.val + "";
        getPaths(result, root, s);
        return result;
    }
    
    public void getPaths(List<String> result, TreeNode root, String s){
        if (root.left == null && root.right == null){
            result.add(s);
            return;
        }
        if(root.left != null){
            getPaths(result, root.left, s + "->" + root.left.val);
        }
        if (root.right != null){
            getPaths(result, root.right, s + "->" + root.right.val);
        }
        
    }
}

File: cpp/257-BinaryTreePaths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        if (!root){
            return res;
        }
        string init = to_string(root->val);
        traverse(root, init, res);
        return res;
    }

    void traverse(TreeNode* root, string curr, vector<string> &res){
        if (!root->left && !root->right){
            res.push_back(curr);
        }
        if (root->left){
            traverse(root->left, curr + "->" + to_string(root->left->val), res);
        }
        if (root->right){
            traverse(root->right, curr + "->" + to_string(root->right->val), res);
        }
    }
};
